### PR TITLE
Add ShedLock provider for Memcached

### PR DESCRIPTION
**Pull request by pinkhello**
Add ShedLock provider for Memcached

### CODE REVIEW

File: providers/memcached/shedlock-provider-memcached/src/main/java/net/javacrumbs/shedlock/provider/memcached/MemcachedLockProvider.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    public MemcachedLockProvider(@NonNull MemcachedClient client, @NonNull String env){
+        this.client = client;
+        this.env = env;
+    }
+
+    /**
+     * set-add-replace
+     * @see https://github.com/memcached/memcached/wiki/BinaryProtocolRevamped#set-add-replace
+     * @param lockConfiguration LockConfiguration
+     * @return
```
[the remark this answers is not in the document]
Comment (pinkhello) [reply providers/memcached/shedlock-provider-memcached/src/main/java/net/javacrumbs/shedlock/provider/memcached/MemcachedLockProvider.java]: done

File: providers/memcached/shedlock-provider-memcached/src/main/java/net/javacrumbs/shedlock/provider/memcached/MemcachedLockProvider.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    public MemcachedLockProvider(@NonNull MemcachedClient client, @NonNull String env){
+        this.client = client;
+        this.env = env;
+    }
+
+    /**
+     * set-add-replace
+     * @see https://github.com/memcached/memcached/wiki/BinaryProtocolRevamped#set-add-replace
+     * @param lockConfiguration LockConfiguration
+     * @return Optional<SimpleLock>
```
[the remark this answers is not in the document]
Comment (pinkhello) [reply providers/memcached/shedlock-provider-memcached/src/main/java/net/javacrumbs/shedlock/provider/memcached/MemcachedLockProvider.java]: del

**Comment by lukas-krecan** [comment]
Hi, thaks a lot. The code looks good (I will look at it in more detail in next few days). My main concern is that memcahced is a cache and if I understand it correctly, it can in theory [evict](https://stackoverflow.com/questions/6868256/memcached-eviction-prior-to-key-expiry/10456364#10456364) the nodes before the expiration time (if the cache gets full). If that happens the lock would be released prematurely whci would be surprising for the library user. I am not memcached expert, how likely is it?

**Comment by pinkhello** [comment]
> Hi, thanks a lot. The code looks good (I will look at it in more detail in next few days). My main concern is that memcahced is a cache and if I understand it correctly, it can in theory [evict](https://stackoverflow.com/questions/6868256/memcached-eviction-prior-to-key-expiry/10456364#10456364) the nodes before the expiration time (if the cache gets full). If that happens, the lock would be released prematurely which would be surprising for the users. I am not memcached expert, how likely is it?



> Hi, thanks a lot. The code looks good (I will look at it in more detail in next few days). My main concern is that memcahced is a cache and if I understand it correctly, it can in theory [evict](https://stackoverflow.com/questions/6868256/memcached-eviction-prior-to-key-expiry/10456364#10456364) the nodes before the expiration time (if the cache gets full). If that happens, the lock would be released prematurely which would be surprising for the users. I am not memcached expert, how likely is it?

Sure!

**Comment by pinkhello** [comment]
> Hi, thanks a lot. The code looks good (I will look at it in more detail in next few days). My main concern is that memcahced is a cache and if I understand it correctly, it can in theory [evict](https://stackoverflow.com/questions/6868256/memcached-eviction-prior-to-key-expiry/10456364#10456364) the nodes before the expiration time (if the cache gets full). If that happens, the lock would be released prematurely which would be surprising for the users. I am not memcached expert, how likely is it?

Sure! That's a problem. We're using a  cluster of memcached servers.

**Pull request closed without merging**